data-processing/pdf/process_pdf.py

```diff
@@ -0,0 +1,236 @@
+from entities.symbols.types import SymbolData, SymbolWithId, SymbolId
+from entities.citations.utils import upload_citations, extract_ngrams, ngram_sim
+from entities.symbols.utils import upload_symbols
+
+class PdfStructureParser(object):
+    def __init__(self, pdf_hash, structure_map):
+        self.pdf_hash = pdf_hash
+        self.structure_map = structure_map
+        self.pages = structure_map['tokens']['pages']
+        self.page_sizes = dict((p['page']['pageNumber'], p['page']) for p in self.pages)
+        self.page_indices = dict((p['page']['pageNumber'], i) for i,p in enumerate(self.pages))
+        self.elements = structure_map['elements']['elementTypes']
+        self.references = None
+
+    def find_token(self, index):
+        for p in self.pages:
+            if (index < len(p['tokens'])):
+                return p['page'], p['tokens'][index]
+            else:
+                index -= len(p['tokens'])
+        return None
+
+
+    def get_text(self,spans):
+        text = []
+        for s in spans:
+            direct = s.get('dehyphenizedText')
+            if direct:
+                text += direct
+            else:
+                for i in range(s['left'], s['right']):
+                    page, token = self.find_token(i)
+                    text.append(token['text'])
+        return ' '.join(text)
+
+    def union(self, bbox1: BoundingBoxModel, bbox2: BoundingBoxModel):
+        if not bbox1:
+            return bbox2
+        x1 = min(bbox1.left,bbox2.left)
+        y1 = min(bbox1.top,bbox2.top)
+        x2 = max(bbox1.left+bbox1.width, bbox2.left+bbox2.width)
+        y2 = max(bbox1.top+bbox1.height, bbox2.top+bbox2.height)
+        return BoundingBoxModel(page = bbox1.page,
+                                left = x1, top = y1, width = x2-x1, height=y2-y1)
+
+    @staticmethod
+    def should_combine(bbox1: BoundingBoxModel, bbox2: BoundingBoxModel):
+        return bbox1.page == bbox2.page and (
+            abs(bbox1.top - bbox2.top) < 4 # Same y-coordinate
+        ) and (
+            abs(bbox2.left - bbox1.left - bbox1.width) < 15 # To the right
+        )
+
+    def get_bounding_boxes(self, spans):
+        bboxes = []
+        for s in spans:
+            bbox = None
+            for i in range(s['left'], s['right']):
+                page, token = self.find_token(i)
+                page_number = page['pageNumber']
+                token_bbox = BoundingBoxModel(
+                    page = page_number,
+                    left = token['x'],
+                    top = token['y'],
+                    width = token['width'],
+                    height=token['height']
+                )
+                if not bbox:
+                    bbox = token_bbox
+                elif self.should_combine(bbox, token_bbox):
+                     bbox = self.union(bbox, token_bbox)
+                else:
+                    bboxes.append(bbox)
+                    bbox = token_bbox
+            bboxes.append(bbox)
```

Comment:
What's this last append for? Will it repeat the work in Line 85?

Reply:
No, because `bbox` is re-assigned after line 85. This line at the end of the loop is needed to add the last bounding box. I should add a check for `None`, though, to handle the empty case.